datafusion/physical-optimizer/src/pushdown_sort.rs

```diff
@@ -0,0 +1,1235 @@
+    }
+
+    fn name(&self) -> &str {
+        "PushdownSort"
+    }
+
+    fn schema_check(&self) -> bool {
+        true
+    }
+}
+
+/// Optimize a SortExec by potentially pushing the sort down to the data source
+fn optimize_sort(sort_exec: &SortExec) -> Result<Transformed<Arc<dyn ExecutionPlan>>> {
+    let sort_input = Arc::clone(sort_exec.input());
+    let required_ordering = sort_exec.expr();
+
+    // First, check if the sort is already satisfied by input ordering
+    if let Some(_input_ordering) = sort_input.output_ordering() {
+        let input_eq_properties = sort_input.equivalence_properties();
+
+        if input_eq_properties.ordering_satisfy(required_ordering.clone())? {
+            return remove_unnecessary_sort(sort_exec, sort_input);
+        }
+    }
+
+    // Try to push the sort requirement down to the data source
+    if let Some(optimized_input) = try_pushdown_sort(&sort_input, required_ordering)? {
+        // Verify that the optimized input satisfies the required ordering
+        if optimized_input
+            .equivalence_properties()
+            .ordering_satisfy(required_ordering.clone())?
+        {
+            return remove_unnecessary_sort(sort_exec, optimized_input);
+        }
+
+        // If not fully satisfied, keep the sort but with optimized input
+        return Ok(Transformed::yes(Arc::new(
+            SortExec::new(required_ordering.clone(), optimized_input)
+                .with_fetch(sort_exec.fetch())
+                .with_preserve_partitioning(sort_exec.preserve_partitioning()),
+        )));
+    }
+
+    Ok(Transformed::no(Arc::new(sort_exec.clone())))
+}
+
+/// Handle the GlobalLimitExec -> SortExec pattern
+fn optimize_limit_sort(
+    limit_exec: &GlobalLimitExec,
+    sort_exec: &SortExec,
+) -> Result<Transformed<Arc<dyn ExecutionPlan>>> {
+    let sort_input = Arc::clone(sort_exec.input());
+    let required_ordering = sort_exec.expr();
+
+    // Check if input is already sorted
+    if let Some(_input_ordering) = sort_input.output_ordering() {
+        let input_eq_properties = sort_input.equivalence_properties();
+        if input_eq_properties.ordering_satisfy(required_ordering.clone())? {
+            // Input is already sorted correctly, remove sort and keep limit
+            return Ok(Transformed::yes(Arc::new(GlobalLimitExec::new(
+                sort_input,
+                limit_exec.skip(),
+                limit_exec.fetch(),
+            ))));
+        }
+    }
+
+    // Try to push down the sort requirement
+    if let Some(optimized_input) = try_pushdown_sort(&sort_input, required_ordering)? {
+        if optimized_input
+            .equivalence_properties()
+            .ordering_satisfy(required_ordering.clone())?
+        {
+            // Successfully pushed down sort, now handle the limit
+            let total_fetch = limit_exec.skip() + limit_exec.fetch().unwrap_or(0);
+
+            // Try to push limit down as well if the source supports it
```

Comment:
I think the current `limit_pushdown` physical optimizer rule can do this. So do we still need to distinguish the sort and limit + sort pattern?